File: OneToOneCallSample/app/src/test/java/com/stringee/apptoappcallsample/stringee/common/AudioDeviceResolverTest.java
package com.stringee.apptoappcallsample.stringee.common;

import static com.stringee.common.StringeeAudioManager.AudioDevice.BLUETOOTH;
import static com.stringee.common.StringeeAudioManager.AudioDevice.EARPIECE;
import static com.stringee.common.StringeeAudioManager.AudioDevice.NONE;
import static com.stringee.common.StringeeAudioManager.AudioDevice.SPEAKER_PHONE;
import static com.stringee.common.StringeeAudioManager.AudioDevice.WIRED_HEADSET;
import static org.junit.Assert.assertEquals;

import com.stringee.common.StringeeAudioManager.AudioDevice;

import org.junit.Test;

import java.util.EnumSet;
import java.util.Set;

public class AudioDeviceResolverTest {
    @Test
    public void voiceDefaultsToEarpieceAndVideoDefaultsToSpeaker() {
        Set<AudioDevice> available = EnumSet.of(EARPIECE, SPEAKER_PHONE);

        assertEquals(EARPIECE, AudioDeviceResolver.resolveBuiltInDevice(false, available));
        assertEquals(SPEAKER_PHONE, AudioDeviceResolver.resolveBuiltInDevice(true, available));
    }

    @Test
    public void activePeripheralIsPreservedDuringInitialSelection() {
        Set<AudioDevice> available = EnumSet.of(
                EARPIECE, SPEAKER_PHONE, BLUETOOTH, WIRED_HEADSET);

        assertEquals(WIRED_HEADSET,
                AudioDeviceResolver.resolveInitialDevice(false, WIRED_HEADSET, available));
        assertEquals(BLUETOOTH,
                AudioDeviceResolver.resolveInitialDevice(true, BLUETOOTH, available));
    }

    @Test
    public void builtInSdkSelectionDoesNotOverrideCallTypeDefault() {
        Set<AudioDevice> available = EnumSet.of(EARPIECE, SPEAKER_PHONE, BLUETOOTH);

        assertEquals(EARPIECE,
                AudioDeviceResolver.resolveInitialDevice(false, SPEAKER_PHONE, available));
        assertEquals(SPEAKER_PHONE,
                AudioDeviceResolver.resolveInitialDevice(true, EARPIECE, available));
        assertEquals(WIRED_HEADSET,
                AudioDeviceResolver.resolveInitialDevice(false, WIRED_HEADSET,
                        EnumSet.of(EARPIECE, SPEAKER_PHONE, WIRED_HEADSET)));
    }

    @Test
    public void returnsNoneWhenNoBuiltInRouteExists() {
        assertEquals(NONE,
                AudioDeviceResolver.resolveBuiltInDevice(false, EnumSet.of(BLUETOOTH)));
    }

    @Test
    public void peripheralRemovalRestoresTheUsersBuiltInChoice() {
        Set<AudioDevice> available = EnumSet.of(EARPIECE, SPEAKER_PHONE);

        assertEquals(EARPIECE,
                AudioDeviceResolver.resolvePreferredBuiltIn(EARPIECE, available));
        assertEquals(SPEAKER_PHONE,
                AudioDeviceResolver.resolvePreferredBuiltIn(SPEAKER_PHONE, available));
    }

    @Test
    public void automaticRoutingDoesNotBecomeAnSdkUserSelection() {
        assertEquals(AudioDeviceResolver.RouteAction.APPLY_BUILT_IN_ONLY,
                AudioDeviceResolver.resolveRouteAction(false, EARPIECE));
        assertEquals(AudioDeviceResolver.RouteAction.OBSERVE_ONLY,
                AudioDeviceResolver.resolveRouteAction(false, BLUETOOTH));
        assertEquals(AudioDeviceResolver.RouteAction.SELECT_DEVICE,
                AudioDeviceResolver.resolveRouteAction(true, SPEAKER_PHONE));
        assertEquals(AudioDeviceResolver.RouteAction.SELECT_DEVICE,
                AudioDeviceResolver.resolveRouteAction(true, WIRED_HEADSET));
    }
}

File: Kotlin-OneToOneCallSample/app/src/test/java/com/stringee/kotlin_onetoonecallsample/stringee/common/AudioDeviceResolverTest.kt
package com.stringee.kotlin_onetoonecallsample.stringee.common

import com.stringee.common.StringeeAudioManager.AudioDevice
import com.stringee.kotlin_onetoonecallsample.stringee.common.AudioDeviceResolver.resolveBuiltInDevice
import com.stringee.kotlin_onetoonecallsample.stringee.common.AudioDeviceResolver.resolveInitialDevice
import com.stringee.kotlin_onetoonecallsample.stringee.common.AudioDeviceResolver.resolvePreferredBuiltIn
import com.stringee.kotlin_onetoonecallsample.stringee.common.AudioDeviceResolver.resolveRouteAction
import org.junit.Assert
import org.junit.Test
import java.util.EnumSet

class AudioDeviceResolverTest {
    @Test
    fun voiceDefaultsToEarpieceAndVideoDefaultsToSpeaker() {
        val available: MutableSet<AudioDevice?>? =
            EnumSet.of<AudioDevice?>(AudioDevice.EARPIECE, AudioDevice.SPEAKER_PHONE)

        Assert.assertEquals(AudioDevice.EARPIECE, resolveBuiltInDevice(false, available))
        Assert.assertEquals(AudioDevice.SPEAKER_PHONE, resolveBuiltInDevice(true, available))
    }

    @Test
    fun activePeripheralIsPreservedDuringInitialSelection() {
        val available: MutableSet<AudioDevice?>? = EnumSet.of<AudioDevice?>(
            AudioDevice.EARPIECE,
            AudioDevice.SPEAKER_PHONE,
            AudioDevice.BLUETOOTH,
            AudioDevice.WIRED_HEADSET
        )

        Assert.assertEquals(
            AudioDevice.WIRED_HEADSET,
            resolveInitialDevice(false, AudioDevice.WIRED_HEADSET, available)
        )
        Assert.assertEquals(
            AudioDevice.BLUETOOTH,
            resolveInitialDevice(true, AudioDevice.BLUETOOTH, available)
        )
    }

    @Test
    fun builtInSdkSelectionDoesNotOverrideCallTypeDefault() {
        val available: MutableSet<AudioDevice?>? = EnumSet.of<AudioDevice?>(
            AudioDevice.EARPIECE,
            AudioDevice.SPEAKER_PHONE,
            AudioDevice.BLUETOOTH
        )

        Assert.assertEquals(
            AudioDevice.EARPIECE,
            resolveInitialDevice(false, AudioDevice.SPEAKER_PHONE, available)
        )
        Assert.assertEquals(
            AudioDevice.SPEAKER_PHONE,
            resolveInitialDevice(true, AudioDevice.EARPIECE, available)
        )
        Assert.assertEquals(
            AudioDevice.WIRED_HEADSET,
            resolveInitialDevice(
                false, AudioDevice.WIRED_HEADSET,
                EnumSet.of<AudioDevice?>(
                    AudioDevice.EARPIECE,
                    AudioDevice.SPEAKER_PHONE,
                    AudioDevice.WIRED_HEADSET
                )
            )
        )
    }

    @Test
    fun returnsNoneWhenNoBuiltInRouteExists() {
        Assert.assertEquals(
            AudioDevice.NONE,
            resolveBuiltInDevice(false, EnumSet.of<AudioDevice?>(AudioDevice.BLUETOOTH))
        )
    }

    @Test
    fun peripheralRemovalRestoresTheUsersBuiltInChoice() {
        val available: MutableSet<AudioDevice?>? =
            EnumSet.of<AudioDevice?>(AudioDevice.EARPIECE, AudioDevice.SPEAKER_PHONE)

        Assert.assertEquals(
            AudioDevice.EARPIECE,
            resolvePreferredBuiltIn(AudioDevice.EARPIECE, available)
        )
        Assert.assertEquals(
            AudioDevice.SPEAKER_PHONE,
            resolvePreferredBuiltIn(AudioDevice.SPEAKER_PHONE, available)
        )
    }

    @Test
    fun automaticRoutingDoesNotBecomeAnSdkUserSelection() {
        Assert.assertEquals(
            AudioDeviceResolver.RouteAction.APPLY_BUILT_IN_ONLY,
            resolveRouteAction(false, AudioDevice.EARPIECE)
        )
        Assert.assertEquals(
            AudioDeviceResolver.RouteAction.OBSERVE_ONLY,
            resolveRouteAction(false, AudioDevice.BLUETOOTH)
        )
        Assert.assertEquals(
            AudioDeviceResolver.RouteAction.SELECT_DEVICE,
            resolveRouteAction(true, AudioDevice.SPEAKER_PHONE)
        )
        Assert.assertEquals(
            AudioDeviceResolver.RouteAction.SELECT_DEVICE,
            resolveRouteAction(true, AudioDevice.WIRED_HEADSET)
        )
    }
}
